[Merged by Bors] - feat: check for PR titles which do not start with "feat" etc.

grunweg opened the pull request:
We ignore draft PRs and PRs with the WIP label.
The set of checks is intentionally kept small, and could be expanded in the
future.

---

More basic version of #16303: given that mathlib's statistics will use the PR kind, let's make sure PR kinds are somewhat appropriate.

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg committed:
feat: check for PR titles which do not start with "feat" etc.

We ignore draft PRs and PRs with the WIP label.
The set of checks is intentionally kept small, and could be expanded in the
future.

Hook up

grunweg committed:
hack: fix CLI parsing

grunweg committed:
Tweak

bryangingechen commented:
I like the idea, but I wonder if this could be made more user-friendly by posting a comment on the PR (maybe just the first time it's triggered, and then cleaning-up afterwards) rather than forcing the user to dig through the CI logs? I think [peter-evans/create-or-update-comment](https://github.com/peter-evans/create-or-update-comment) which we use in the merge conflict workflow can handle this sort of thing.

bryangingechen commented:
(I can add a commit doing that tomorrow if you'd rather not do it.)

eric-wieser commented on Mathlib/Tactic/Linter/ValidatePRTitle.lean:
```diff
@@ -0,0 +1,51 @@
+  let knownKinds := ["feat", "chore", "perf", "refactor", "style", "fix", "doc", "test"]
+  let mut isFine := false
+  for kind in knownKinds do
+    if type.startsWith kind then isFine := true
```
This will allow `feature(foo):`, which presumably we don't want; perhaps we should also split at `(`.

You might be able to use some of the `Parser` machinery to write the parser here, which could be fun.

grunweg replied:
I agree that these checks are far too lenient. I tried to be conservative within linting everything, but I'm happy to make it stricter! :-)

grunweg commented:
Posting a comment would be nice; and your help is very welcome here!

bryangingechen committed:
add / update comment if check fails

Adapted from https://github.com/thehanimo/pr-title-checker?tab=readme-ov-file#using-output-result

bryangingechen committed:
Update Mathlib/Tactic/Linter/ValidatePRTitle.lean

bryangingechen committed:
Merge branch 'master' into badtitles-basic

bryangingechen committed:
incorporate errors into message, put details into details

bryangingechen committed:
Merge remote-tracking branch 'upstream/master' into badtitles-basic

bryangingechen committed:
add README, various fixes

bryangingechen committed:
mk_build_yml

bryangingechen committed:
put errors in code block

bryangingechen committed:
Apply suggestion from @bryangingechen

grunweg commented on .github/build.in.yml:
```diff
@@ -577,6 +577,69 @@ jobs:
+
+            - use imperative, present tense: "change" not "changed" nor "changes"
+            - do not capitalize the first letter
+            - no dot(.) at the end
```
I'm not sure if the flashing light is necessary --- but this is much better already, thanks!

bryangingechen committed:
fix deprecations

grunweg committed:
Rewrite PR title validation to use parsec (more)

grunweg commented:
A welcome tweak would be to make the parser also extract the main PR title. (I don't know how to do so quickly, and would be happy to take it from there. This can be in a follow-up!)

grunweg committed:
Lint

adomani commented on Mathlib/Tactic/Linter/ValidatePRTitle.lean:
```diff
@@ -0,0 +1,91 @@
+  Prod.mk
+    <$> (["feat", "chore", "perf", "refactor", "style", "fix", "doc", "test", "ci"].firstM pstring)
+    <*> (
+      (skipString "(" *> some <$> manyChars (notFollowedBy (skipString "):") *> any)
```
```suggestion
      (skipString "(" *> notFollowedBy (skipString "(") *>
        some <$> manyChars (notFollowedBy (skipString "):") *> any)
```
this should also check that `feat((x...` does not happen.

adomani commented on Mathlib/Tactic/Linter/ValidatePRTitle.lean:
```diff
@@ -0,0 +1,91 @@
+-/
+def validateTitle (title : String) : Array String := Id.run do
+  -- The title should be of the form "abbrev: main title" or "abbrev(scope): main title".
+  -- We use the parser above to extract abbrev, scope and main title,
```
```suggestion
  -- We use the parser above to extract abbrev and scope ignoring the main title,
```
If I understand correctly, this is what the parser currently does, right?

adomani commented on Mathlib/Tactic/Linter/ValidatePRTitle.lean:
```diff
@@ -0,0 +1,91 @@
+    let knownKinds := ["feat", "chore", "perf", "refactor", "style", "fix", "doc", "test", "ci"]
+    let mut isFine := false
+    for k in knownKinds do
+      if kind.startsWith k then isFine := true
```
```suggestion
      isFine := isFine || kind.startsWith k
```

grunweg committed:
Apply suggestions from code review

Co-authored-by: damiano <adomani@gmail.com>

bryangingechen committed:
Merge remote-tracking branch 'upstream/master' into badtitles-basic

bryangingechen committed:
move steps to new workflow

bryangingechen committed:
pull_request_target, permissions, job condition

adomani commented on .github/workflows/check_pr_titles.yaml:
```diff
@@ -0,0 +1,118 @@
+      - name: Checkout
+        uses: actions/checkout@08c6903cd8c0fde910a37f88322edcfb5dd907a8 # v5.0.0
+        with:
+          ref: master
```
```suggestion
          ref: master
```
should this point to the branch with the PR?

adomani commented on .github/workflows/check_pr_titles.yaml:
```diff
@@ -0,0 +1,112 @@
+        with:
+          auto-config: false
+          use-github-cache: false
+          use-mathlib-cache: false
```
I agree: the mathlib dependencies of the `lake exe check-title-labels` are really tiny.

bryangingechen committed:
Update .github/workflows/check_pr_titles.yaml

Co-authored-by: damiano <adomani@gmail.com>

adomani commented:
This looks good to me!  I do not know if you want to do some further tweaks/tests, so I am going to delegate a list of authors: let's see if I get this right!

bors d=@grunweg,@bryangingechen

grunweg committed:
Merge branch 'master' into badtitles-basic

grunweg committed:
Fix

grunweg committed:
Fix merge

bryangingechen committed:
more fixes

bryangingechen commented:
bors r+
